tempodb/encoding/vparquet4/block_traceql.go

```diff
@@ -0,0 +1,2403 @@
+		at, _ = i.Next()
+	}
+
+	return at, nil
+}
+
+func (i *bridgeIterator) Close() {
+	i.iter.Close()
+	parquetquery.DefaultPool.Release(i.at)
+}
+
+// confirm rebatchIterator implements parquetquery.Iterator
+var _ parquetquery.Iterator = (*rebatchIterator)(nil)
+
+// rebatchIterator either passes spansets through directly OR rebatches them based on metadata
+// in OtherEntries
+type rebatchIterator struct {
+	iter      parquetquery.Iterator
+	at        *parquetquery.IteratorResult
+	nextSpans []*span
+}
+
+func newRebatchIterator(iter parquetquery.Iterator) *rebatchIterator {
+	return &rebatchIterator{
+		iter: iter,
+		at:   parquetquery.DefaultPool.Get(),
+	}
+}
+
+func (i *rebatchIterator) String() string {
+	return fmt.Sprintf("rebatchIterator: \n\t%s", util.TabOut(i.iter))
+}
+
+// Next has to handle two different style results. First is an initial set of spans
+// that does not have a callback spanset. These can be passed directly through.
+// Second is a set of spans that have spansets imposed by the callback (i.e. for grouping)
+// these must be regrouped into the callback spansets
+func (i *rebatchIterator) Next() (*parquetquery.IteratorResult, error) {
+	for {
+		// see if we have a queue
+		res := i.resultFromNextSpans()
+		if res != nil {
+			return res, nil
+		}
+
+		// check the iterator for anything
+		res, err := i.iter.Next()
+		if err != nil {
+			return nil, err
+		}
+		if res == nil {
+			return nil, nil
+		}
+
+		// get the spanset and see if we should pass it through or buffer for rebatching
+		iface := res.OtherValueFromKey(otherEntrySpansetKey)
+		if iface == nil {
+			return nil, fmt.Errorf("engine assumption broken: spanset not found in other entries in rebatch")
+		}
+		ss, ok := iface.(*traceql.Spanset)
+		if !ok {
+			return nil, fmt.Errorf("engine assumption broken: spanset is not of type *traceql.Spanset in rebatch")
+		}
+
+		// if this has no call back spanset just pass it on
+		if len(ss.Spans) > 0 && ss.Spans[0].(*span).cbSpanset == nil {
+			return res, nil
+		}
+
+		// dump all spans into our buffer
+		for _, s := range ss.Spans {
+			sp := s.(*span)
+			if !sp.cbSpansetFinal {
+				continue
+			}
+
+			// copy trace level data from the current iteration spanset into the rebatch spanset. only do this if
+			// we don't have current data
+			if sp.cbSpanset.DurationNanos == 0 {
+				sp.cbSpanset.DurationNanos = ss.DurationNanos
+			}
+			if len(sp.cbSpanset.TraceID) == 0 {
+				sp.cbSpanset.TraceID = ss.TraceID
+			}
+			if len(sp.cbSpanset.RootSpanName) == 0 {
+				sp.cbSpanset.RootSpanName = ss.RootSpanName
+			}
+			if len(sp.cbSpanset.RootServiceName) == 0 {
+				sp.cbSpanset.RootServiceName = ss.RootServiceName
+			}
+			if sp.cbSpanset.StartTimeUnixNanos == 0 {
+				sp.cbSpanset.StartTimeUnixNanos = ss.StartTimeUnixNanos
+			}
+			if len(sp.cbSpanset.ServiceStats) == 0 {
+				sp.cbSpanset.ServiceStats = map[string]traceql.ServiceStats{}
```

Comment:
If the spanset is allocated in
https://github.com/grafana/tempo/blob/148ba6dbb980f27fff57073a534a44a698f3c02d/pkg/traceql/ast_execute.go#L36
 we need to allocate the `ServiceStats` map here, otherwise it'll get allocated in the keepgroup afaics.

Addressed in https://github.com/stoewer/tempo/pull/5